### PR TITLE
refactor(lua): Use API objects for Lua stdlib

### PR DESCRIPTION
**Pull request by lewis6991**
Instead of creating hand-crafted Lua bindings for the Lua stdlib, now we
re-leverage the same generators for creating the API functions.

This automates some boilerplate, like argument validation, as well as
avoiding needing to use Lua's stack API which can be awkward.

### CODE REVIEW

File: src/nvim/generators/gen_lua_dispatch.lua [hunk 1]
```diff
@@ -0,0 +1,324 @@
+-- we need at least 4 arguments since the last two are output files
+if arg[1] == '--help' then
+  print('Usage: genmsgpack.lua args')
```
Comment (clason) [inline src/nvim/generators/gen_lua_dispatch.lua:3]: copy-pasta? ;)

Comment (lewis6991) [reply src/nvim/generators/gen_lua_dispatch.lua:3]: Yep, from `gen_api_dispatch.lua` from which it was also copy-pasta'd 😆

**Comment by justinmk** [comment]
> This automates some boilerplate, like argument validation, as well as avoiding needing to use Lua's stack API which can be awkward.

Avoiding boilerplate is good. Better validation without boilerplate is good. But is that shown in this PR yet?

Avoiding Lua API might not be needed if we don't plan to add a lot of C-level Lua functions.

**Comment by lewis6991** [comment]
> Avoiding boilerplate is good. Better validation without boilerplate is good. But is that shown in this PR yet? 

No.

This PR is a result of a discussion on matrix where bfredl said he partly implemented `nvim_set_decoration_provider` in the API because it was more convenient to use the API objects, despite the fact the function is Lua only. We agreed Lua only functions should be kept out of the API. (see #19164)

> The new stricmp impl still looks roughly the same.

`stricmp` is a very simple function. I've simply refactored this one as an example. For a more complicated API, especially ones that include table arguments, the benefit should be more clear as processing tables in Lua-C is quite tedious.

> Avoiding Lua API might not be needed if we don't plan to add a lot of C-level Lua functions.

Agree. This PR is predicated on that will add more Lua only functions. I won't push for this PR if we don't need it. There is also no priority to get this in soon, but the foundations are now ready if we do decide this is worth it.

Using the API object style functions gives us quite a bit of flexibility in the future:
- These can be exposed to libnvim.
- Deciding whether a function goes in the API or Lua will just be a matter of naming the function and placing it in the right file. The implementation stays exactly the same.

**Pull request closed without merging**